### ISSUE TITLE
Make Fail eaa6ca5

### ISSUE THREAD

**Issue by JackJollimore** [ggerganov/llama.cpp#2015]
Hi,

make fails to build with eaa6ca5

here's the error:
```
~/ollama> make
I llama.cpp build info:
I UNAME_S:  Linux
I UNAME_P:  unknown
I UNAME_M:  aarch64
I CFLAGS:   -I.              -O3 -std=c11   -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wdouble-promotion -Wshadow -Wstrict-prototypes -Wpointer-arith -pthread -DGGML_USE_K_QUANTS -mcpu=native
I CXXFLAGS: -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS -mcpu=native
I LDFLAGS:
I CC:       clang version 16.0.6
I CXX:      clang version 16.0.6

fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
cc  -I.              -O3 -std=c11   -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wdouble-promotion -Wshadow -Wstrict-prototypes -Wpointer-arith -pthread -DGGML_USE_K_QUANTS -mcpu=native   -c ggml.c -o ggml.o
ggml.c:16644:14: error: call to undeclared function 'pthread_setaffinity_np'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    int rv = pthread_setaffinity_np(pthread_self(), setsize, cpus);
             ^
ggml.c:16666:14: error: call to undeclared function 'pthread_setaffinity_np'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    int rv = pthread_setaffinity_np(pthread_self(), setsize, cpus);
             ^
2 errors generated.
make: *** [Makefile:263: ggml.o] Error 1
```

I also tried cmake and received the same error messages.

Thank you.

**Comment by Green-Sky** [comment]
> make fails to build with [eaa6ca5](https://github.com/ggerganov/llama.cpp/commit/eaa6ca5a61b8c9501df9ebe3d264f45b75a5f8aa)

you probably meant this commit https://github.com/ggerganov/llama.cpp/commit/b853d456018b10820686362af41b2f2f75f1eec6

looks like `pthread_*_np` are gnu extensions and might not be available everywhere. But you are on linux, so it should be available. I can see you are on arm, what device/os are you using?

**Comment by JackJollimore** [comment]
> > make fails to build with [eaa6ca5](https://github.com/ggerganov/llama.cpp/commit/eaa6ca5a61b8c9501df9ebe3d264f45b75a5f8aa)
> 
> you probably meant this commit [b853d45](https://github.com/ggerganov/llama.cpp/commit/b853d456018b10820686362af41b2f2f75f1eec6)
> 

Possibly? I figured the issue would resolve itself by downloading the latest release, but I'm wrong. I tried to build master-0be54f7 and have the same error.

> looks like `pthread_*_np` are gnu extensions and might not be available everywhere. But you are on linux, so it should be available. I can see you are on arm, what device/os are you using?

I use an Android device, Samsung s10+ through Termux.

**Comment by Green-Sky** [comment]
did a little digging and SO pointed me to this: https://android.googlesource.com/platform/bionic/+/master/docs/status.md which lists supported libc functionallity by bionic, androids libc implementation.

the fix would be to test for `__BIONIC__` and don't use setaffinity when it is define.

i opened an pr to fix this #2020

**Comment by JackJollimore** [comment]
> did a little digging and SO pointed me to this: https://android.googlesource.com/platform/bionic/+/master/docs/status.md which lists supported libc functionallity by bionic, androids libc implementation.
> 
> the fix would be to test for `__BIONIC__` and don't use setaffinity when it is define.
> 
> i opened a pr to fix this #2020

Thanks for your response. I'm going to try to pull this, but it might take me a minute to figure out how to do so, so I'll report back as soon as possible.

**Comment by JackJollimore** [comment]
I didn't figure out git, instead I copy/pasted the solution into my ggml.c file which worked.

Your solution allows me to make, thank you!

I'll close for now trusting your solution will be implemented going forward.
